### DISCUSSION TITLE
Pressure Transducer Connection

### DISCUSSION THREAD

**Discussion by Hect0rG** [tagyoureit/nodejs-poolController#416] [Q&A]
I am trying to get my pressure transducer operating properly. It is a 5V logic transducer so as suggested earlier in gitter I when through a level shifter to get my voltage down from 5v to 3v3. I currently treat it like I do all my other temperature sensors but I am not sure if I need to put it through the voltage divider or not. 

I currently have it set up this way on REM.
![image](https://user-images.githubusercontent.com/6274208/163692684-e29c5074-a53c-4b74-b2f8-f497e39cad7f.png)

![image](https://user-images.githubusercontent.com/6274208/163692507-f17d6f26-31cd-470d-b656-6db720d04493.png)

![image](https://user-images.githubusercontent.com/6274208/163692584-b997bafe-0fa2-44a5-b4ac-0d85488ea00e.png)

![image](https://user-images.githubusercontent.com/6274208/163692719-da582356-4320-42e3-89f2-f75ed41e669c.png)

I had to bring the PSI per Volt way down to 9 from the preset 30 to get it to measure what I actually have showing in my pressure sensor. I am thinking that this is due to it going through the voltage divider after it is already 3v3. Before I try this I just wanted to get some input from the community.

**Comment by rstrouse** [comment]
What do you have the gain set to for the ADC.  Remember the voltage through the level shifter is still only 3.3v (2.048v).

**Reply by Hect0rG** [reply]
Yes that is how I have it set now. 

![image](https://user-images.githubusercontent.com/6274208/163695286-8217cd1c-eacf-43dc-9dda-bfa9213a0fd1.png)

**Comment by rstrouse** [comment]
Which transducer do you have?  Specifically, what is its range?

**Comment by rstrouse** [comment]
I found a video online that gives a pretty good description of how to wire the pressure transducer with a level shifter.  There should not be a voltage divider for this since that is all inside the pressure transducer.  You will find the transducer wiring at 10:45 on the video.
https://www.youtube.com/watch?v=wJgyszOSoQU

**Reply by Hect0rG** [reply]
Ok great thanks for that link, I'll take a look at that video when I get a moment. In the meantime the transducer I have has a range of 0 - 80 psi. This is the one https://amzn.to/3EnYqkn.

**Reply by Hect0rG** [reply]
Ok so I have to remove the voltage divider. I'll fix that when I get a chance and report back. Thanks for the video. He explains it well. I took a screenshot of the connections from the video in case it helps someone in the future.

![Screenshot_20220417-132452_Brave](https://user-images.githubusercontent.com/6274208/163725608-c37e8ebc-c2b6-41fc-a2fe-47ba9da2ada0.jpg)

**Reply by orrious** [reply]
I was searching past posts and want to give some background to this schematic.  Level shifters will convert signals, bi-directionally, and only digitally.  What Hect0rG has done is changed the voltage that the entire ADC is running at.  Rather than running at the RPI 3.3v, we are level shifting the I2C bus to 5v and then run the ADC at 5v.

**Comment by Hect0rG** [comment]
Hi folks, I was catching up on this post I had some time ago because I never got it to work correctly. I was reading some other posts on the internet and most suggested running the ADC at 3.3v so I tried setting it up as shown below. The two temp probes are set up normally on the ADC at 3v3. These are working correctly. The pressure sensor is being powered at 5V but going through a logic shifter that brings it down to 3v3 for the ADC. In my mind this should work however when I set it up, the pressure readings are incorrect. When everything is off I am reading 7psi and when the pressure is at 12PSI at the gauge it is showing 35psi on the GUI. 

![image](https://user-images.githubusercontent.com/6274208/185602982-d35b5dc0-1f44-4e4b-bc15-d8855d9d8369.png)

**Reply by orrious** [reply]
The level shifter is digital, not analog. You need to run your ADC at 5v and use the level shifter to shift the I2C voltage from the ADC from 5v to 3.3. Long story short, swap your ADC and level shifter positions. 

If you use the same ADC for your temp probes, run the temp probes at 5v too

**Reply by Hect0rG** [reply]
Got it Thanks. Oversight on my part. That would explain the erroneous readings. I'll give that a try when I get some time.

**Reply by Hect0rG** [reply]
Before I re-wire this thing up, would you mind taking a look at this?  I have now moved the level shifter to be between the Pi and the ADC . I then moved the temp sensors to get powered by the 5V pin instead of the original 3v3 and connected them to the level shifter on A0 and A1. I have looked up how to connect the temp sensors to 5V and in some diagrams, I see the pull-up resistor connected to 3v3 and in others directly to the 5V as shown here. I believe this will work correctly now but just wanted to pass it by you before I take out the soldering gun.

![image](https://user-images.githubusercontent.com/6274208/185815398-73fe1a8f-8c82-48ae-b958-1b010aea7d6e.png)

**Reply by orrious** [reply]
That looks like it should do it.

**Reply by Hect0rG** [reply]
Thanks, I tried today and it looks like my pressure sensor may now be working but my two temp sensors are not. They are both reporting 0 degrees Fahrenheit. I will make sure I didn't overlook anything in the connections soon and report back.

**Reply by rstrouse** [reply]
@Hect0rG are you using DS18B20 sensors?  If those are digital temp sensors, they should be connected to the 1 wire interface not through an analog to digital converter.

**Reply by Hect0rG** [reply]
I am using the 10k Thermistors. They were working before, it looks like I have a false contact somewhere so I am troubleshooting it.

**Reply by Hect0rG** [reply]
I got them to read, my ADC had a bad contact but both are reading the same exact value which I find very odd. It's about 90 degrees out here and the temp is showing about half so something is wrong for sure. I might tone it out for continuity again tomorrow to make sure I didn't create a short or something.

![image](https://user-images.githubusercontent.com/6274208/187006707-46c3deff-4114-4142-ad28-b5f6e215fca9.png)

**Reply by Hect0rG** [reply]
Alright folks I am reaching out for help again. I can't figure this out. I have created a new board and switched out my ADC, and level shifter. I have confirmed all connections and I removed my Argon hat that I had cooling the pi (it seemed to interfere with my I2C) to rule it out. I am convinced that there is something wrong with my schematic but I can't figure it out. I continue getting the same value for both Temp sensors. The Pressure for the filter seems to be working properly but both temp sensors are reading the same exact value which I know is wrong. The resistors I am using are 4.7k in case that helps. Sorry in advance about the excessive amount of screenshots, just want to provide as much detail as possible. As you can see below, both temp sensors are measuring the same exact value even though one sensor is in the air and the other is in the pipe touching the water. The temperature doesn't seem to fluctuate at all. Could the thermistors somehow have shorted out maybe or is it my desing?

I have pulled the latest source and installed it so running the latest and greatest code. Here is how I currently have it all configured.

![image](https://user-images.githubusercontent.com/6274208/187572472-96110331-5cc7-4433-b157-2a677ddf28cc.png)

![image](https://user-images.githubusercontent.com/6274208/187572527-a8de2348-db0a-4496-ac7d-c58b743081ea.png)

![image](https://user-images.githubusercontent.com/6274208/187572587-c369f910-9bf4-4a93-9448-994fa35a8ad9.png)

Both Temp sensors are set up the same 
![image](https://user-images.githubusercontent.com/6274208/187572613-90e5454d-7011-4806-a424-09d38564bfbf.png)

![image](https://user-images.githubusercontent.com/6274208/187572689-4a6290c2-07ef-41c6-8d9b-3e47a1de8b8c.png)

![image](https://user-images.githubusercontent.com/6274208/187572729-b74a13bc-71ca-4256-8f88-c9ae4928731d.png)

![image](https://user-images.githubusercontent.com/6274208/187572778-b48eff97-bbbb-46c0-997b-10b2b437dd92.png)

![image](https://user-images.githubusercontent.com/6274208/187572827-188eb952-fe9a-4c7a-a767-05e163b5279f.png)

![image](https://user-images.githubusercontent.com/6274208/187572877-0c61f5b3-ec13-4075-ab12-12a508b09758.png)

![image](https://user-images.githubusercontent.com/6274208/187572937-41075521-867d-48f5-8fd3-50f28c915caa.png)

![image](https://user-images.githubusercontent.com/6274208/187572989-0f39abc2-1aa9-44e8-83de-ced98d1d5fab.png)

![image](https://user-images.githubusercontent.com/6274208/187573025-5f283769-40da-4b94-9b4e-b64dd79e8473.png)

![image](https://user-images.githubusercontent.com/6274208/187573062-e7ba10d1-c435-488b-b0fd-7a1c8f4d7697.png)

![image](https://user-images.githubusercontent.com/6274208/187573084-f6a3fac1-2631-4b18-b707-b13325710cc6.png)

![image](https://user-images.githubusercontent.com/6274208/187573716-c07e82f3-e675-4194-b814-95eca21d0d1a.png)

**Reply by rstrouse** [reply]
If you are using 10k temperature sensors you need to be using 10k resistors for your voltage divider.  That is a problem.

Hook them up like this.
https://github.com/rstrouse/relayEquipmentManager/wiki/Temperature-Sensors

**Reply by Hect0rG** [reply]
Thanks for that, I dont know how many times I have looked at that drawing and never noticed the resistor value. Ill give that a try tomorrow.

**Reply by Hect0rG** [reply]
ok I swapped out the resistors to 10k. They are connected the same way except they are on the 5V leg instead of 3v3. Unfortunately, it didn't fix anything and voltages are still the same on both channels.

**Reply by rstrouse** [reply]
Try increasing the gain on the channel.

**Reply by Hect0rG** [reply]
Reducing the gain actually gets me closer to the real temp but it's off. It is currently 79 degrees outside and pool controller is reporting 57 with a power gain of 2.48. If I set it to 4.096 the temp drops to 45 degrees. Dropping the gain to 1.024 makes the temp go up to 111.64 degrees. I checked the thermistor and it is just under 10k reading on the impedance meter so I don't think that is the issue. One question is since I am feeding the pi from the level shifter, I assumed my vref should remain at 3.3 but when I leave the gain at 2.48 and increase the Vref to 5, it seems to get a bit closer (still wrong 92.39) to the real value. I still don't think that is the issue since both thermistors one in the air and one in the water are reporting the same exact value. I am now starting to suspect the pi's I2C bus is faulty.

**Reply by tagyoureit** [reply]
@Hect0rG I looked _really_ closely at your diagram and you have the DS18B20 temperature sensors.  Those are 1-wire temperature sensors (which we do work with) but they are not supposed to be used with ADC's.  [1-wire Wiki](https://github.com/rstrouse/relayEquipmentManager/wiki/1-Wire-Devices).

**Reply by Hect0rG** [reply]
Ok I goofed, that drawing is wrong. I am using the 10k thermistor that only have 2 legs. I will update the drawing later today. I have the thermistor wired the way you showed here https://github.com/rstrouse/relayEquipmentManager/wiki/Temperature-Sensors

**Reply by Hect0rG** [reply]
@tagyoureit  sorry about that, here is the updated drawing.

![image](https://user-images.githubusercontent.com/6274208/188033614-0cc4276b-f335-4c49-8703-1903da609265.png)

**Reply by orrious** [reply]
Change the polarity of the voltage divider network.

5v+ -> Thermistor
Signal -> Thermistor and resistor
GND -> Resistor

**Reply by Hect0rG** [reply]
ok I will give that a try later tonight. Thanks

**Reply by Hect0rG** [reply]
Ok, I gave that a try and the odd part is that the values are identical to before I made the change. This is just odd because by swapping the polarity those values definitely had to change and I do not see any evidence of that. Is there any way to see the values that are being sent by I2C from the pi before it is supposed to hit the controller?

![image](https://user-images.githubusercontent.com/6274208/188286870-059ba49f-80be-4a31-9bc8-598532450875.png)

This is how I have it wired now.
![image](https://user-images.githubusercontent.com/6274208/188286939-5cb9211d-b2f4-415f-8992-fb4082a979dd.png)

**Reply by tagyoureit** [reply]
Check out this [Adafruit tutorial](https://learn.adafruit.com/thermistor/using-a-thermistor).  It's for an Arduino, but it walks you through how you can setup and read the voltage signals with a simple multimeter and shows you how to hook it up to a analog input.  You'll do the same except your input will be to the ADC, not the board pin.

![image](https://user-images.githubusercontent.com/1646840/188299634-b65dae93-998b-4488-a1cf-bf87a8a77e61.png)

In the diagram here, you see they have the 10k resistor between the 5v and data lines.  In your diagram above you should have the resistor between the 5v and data as well (red/gray and red/yellow wires for each thermistor on one leg).  Yours is currently setup with the resistor on the ground which you do not want.

**Reply by Hect0rG** [reply]
Ok I have an arduino and a breadboard and will set it up to read from there. That way I can verify the thermistors are not the problem. I also agree that the way I have seen the voltage divider set up is how I originally had it with the resistor to the +5. I have been messing with voltage divider calculators to see what my outputs should be and they also line up with 5v on the resistor side. Anyways thanks for the help. I'll report back as soon as I can.

**Reply by Hect0rG** [reply]
ok, I have tested on a breadboard, and the way shown above works as expected. I rewired my circuit on the circuit board to match how I had it before. 

![image](https://user-images.githubusercontent.com/6274208/188457431-fc120b9a-95fd-45b2-801d-5bf9d4edd98f.png)

I then connected the A0 pin to an Arduino Uno I was testing with and sure enough, the circuit is working as expected. 

The sketch I used is a compilation of a few sketches I found while researching. I pasted it here for reference.

```
// SPDX-FileCopyrightText: 2011 Limor Fried/ladyada for Adafruit Industries
//
// SPDX-License-Identifier: MIT

// thermistor-1.ino Simple test program for a thermistor for Adafruit Learning System
// https://learn.adafruit.com/thermistor/using-a-thermistor by Limor Fried, Adafruit Industries
// MIT License - please keep the attribution and consider buying parts from Adafruit

// the value of the 'other' resistor
#define SERIESRESISTOR 10000    
 
 // The beta coefficient of the thermistor (usually 3000-4000)
#define BCOEFFICIENT 3950

// temp. for nominal resistance (almost always 25 C)
#define TEMPERATURENOMINAL 25   

// What pin to connect the sensor to
#define THERMISTORPIN A0 


float logR2, R2, T, Tc, Tf;
float c1 = 1.009249522e-03, c2 = 2.378405444e-04, c3 = 2.019202697e-07;


void setup(void) {
  Serial.begin(9600);
}
 
void loop(void) {
  float reading, readingx;
 
  reading = analogRead(THERMISTORPIN);
 
  Serial.print("Analog reading "); 
  Serial.println(reading);

  // convert the value to resistance
  readingx = 1023 / reading - 1;
  readingx = SERIESRESISTOR / readingx;
  Serial.print("Thermistor resistance "); 
  Serial.println(readingx);
  
  float steinhart;
  steinhart = readingx / SERIESRESISTOR;     // (R/Ro)
  steinhart = log(steinhart);                  // ln(R/Ro)
  steinhart /= BCOEFFICIENT;                   // 1/B * ln(R/Ro)
  steinhart += 1.0 / (TEMPERATURENOMINAL + 273.15); // + (1/To)
  steinhart = 1.0 / steinhart;                 // Invert
  steinhart -= 273.15;                         // convert absolute temp to C
  Tf = (steinhart * 9.0)/ 5.0 + 32.0; 
  
  Serial.print("Temperature "); 
  Serial.print(steinhart);
  Serial.println(" *C");

  Serial.print("Temperature: "); 
  Serial.print(Tf);
  Serial.println(" *F");
  Serial.print('\n');
 
  delay(1000);
}
```


My Arduino sketch output was as follows while I was holding the thermistor firmly in my hand (should be about 92 *F). Once I let go, the temperature goes back down to 76 *F since I am testing inside my home. 

Analog reading 404.00
Thermistor resistance 6526.66
Temperature 34.92 *C
Temperature: 94.86 *F

Once I let go of the probe and allow it to cool down, the reading is as follows.

Analog reading 515.00
Thermistor resistance 10137.79
Temperature 24.69 *C
Temperature: 76.45 *F

I then set up the following sketch on the Uno to test that the I2C readings are accurate and coming in with no issue. 


```
#include <Wire.h>
#include <Adafruit_ADS1X15.h>
 
Adafruit_ADS1115 ads;
const float multiplier = 0.1875F;
 
void setup(void) 
{
  Serial.begin(9600);
 
  // Descomentar el que interese
  // ads.setGain(GAIN_TWOTHIRDS);  +/- 6.144V  1 bit = 0.1875mV (default)
  // ads.setGain(GAIN_ONE);        +/- 4.096V  1 bit = 0.125mV
  // ads.setGain(GAIN_TWO);        +/- 2.048V  1 bit = 0.0625mV
  // ads.setGain(GAIN_FOUR);       +/- 1.024V  1 bit = 0.03125mV
  // ads.setGain(GAIN_EIGHT);      +/- 0.512V  1 bit = 0.015625mV
  // ads.setGain(GAIN_SIXTEEN);    +/- 0.256V  1 bit = 0.0078125mV 
  ads.begin();
}
 
void loop(void) 
{
  int16_t adc0, adc1, adc2, adc3;
 
  
  adc0 = ads.readADC_SingleEnded(0);
  adc1 = ads.readADC_SingleEnded(1);
  adc2 = ads.readADC_SingleEnded(2);
  adc3 = ads.readADC_SingleEnded(3);
  Serial.print("AIN0: "); Serial.println(adc0 * multiplier);
  Serial.print("AIN1: "); Serial.println(adc1 * multiplier);
  Serial.print("AIN2: "); Serial.println(adc2 * multiplier);
  Serial.print("AIN3: "); Serial.println(adc3 * multiplier);
  Serial.println(" ");
  
  delay(1000);
}
```


----------------------------------------------------------------------------------------------------------------------------------------------------
**Measuring at the ADC prior to the Level Shifter**
If I measure at the ADC prior to the level shifter when my hand is wrapped around the thermistor, the value is as shown below. My hands made it roughly 92 *F. 

AIN0: 2062.69
AIN1: 4905.75
AIN2: 4905.19
AIN3: 549.00

After letting it go for a few minutes the analog value prior to the level shifter climbs. It's about 76 degrees give or take. 

AIN0: 2310.75
AIN1: 4899.00
AIN2: 4897.50
AIN3: 546.75

----------------------------------------------------------------------------------------------------------------------------------------------------
**After the Level Shifter**
When I grip the thermistor to make it hot (approx 92 *F), the readings are as follows.  The analog value goes down as it gets hotter. This seems correct and is after the level shifter shifts down to 3v3 levels.

AIN0: 1938.75
AIN1: 4902.75
AIN2: 4899.37
AIN3: 557.44     

After letting it go for a few minutes the analog value climbs. It's about 76 degrees give or take. 

AIN0: 2386.12
AIN1: 4907.62
AIN2: 4904.25
AIN3: 552.75


So in conclusion, I am not sure how these values are interpreted by nodejs pool controller and REM but it seems to me that this should work leading me back to the issue being with the RaspberryPi's I2C. Is there any way for me to tell what those values should return as a temp in REM? is 1938.75 ~ 92*F and 2386.12 ~ 76 *F? It seems that the level shifter isn't doing much here could that be part of my problem?

**Reply by tagyoureit** [reply]
The thermistor table we use is pretty standard: https://github.com/rstrouse/relayEquipmentManager/blob/master/devices/thermistorTable.json

We typically use a PGA of 2.048v because we will be closest to a 3v3v input if that's what you are using.  This will affect your raw ADC reading.  Try setting that in your code.

And the formula we use to go from adc to voltage which is displayed on the ADS1115 device: 
                    // voltage = value / max * pga
                    // 22,127 = value
                    // 65356 = max
                    // 2.048 = pga
                    // 22,127 / 65536 * 2.048 = 0.69v

But the ADC value is what is passed to the generic 10k temp device.  Resistance is calculated as:
(10000 * adcvalue) / (max adc value- adc value)
eg 10000 * 22127 / (65536 - 22127) = 5097 ohms
Looking this up on the thermistor table it translates to ~105.97 deg F.

These screenshots were taken at different times so ignore the fact the numbers don't match.  
![2022-09-05_14-31-09](https://user-images.githubusercontent.com/1646840/188514098-d68b0210-a045-40da-aaf3-a35b4119a873.png)

**Reply by Hect0rG** [reply]
Thanks for that info. Ill give that a try.

**Reply by Hect0rG** [reply]
I found a python program that would let me read the I2C values and modified it to display the raw value, voltage, and resistance value. I installed node-js and REM on a different pi just to make sure it wasn't a problem with the pi itself. For some reason, they aren't matching what nodejs pool controller is showing. 

Here is the code

```
# SPDX-FileCopyrightText: 2021 ladyada for Adafruit Industries
# SPDX-License-Identifier: MIT

import time
import board
import busio
import adafruit_ads1x15.ads1015 as ADS
from adafruit_ads1x15.analog_in import AnalogIn


# Create the I2C bus
i2c = busio.I2C(board.SCL, board.SDA)

# Create the ADC object using the I2C bus
ads = ADS.ADS1015(i2c)


max_adc = 65356
pga = 2.048


# Create single-ended input on each channel
chan0 = AnalogIn(ads, ADS.P0)
chan1 = AnalogIn(ads, ADS.P1)
chan2 = AnalogIn(ads, ADS.P2)
chan3 = AnalogIn(ads, ADS.P3)

rchan0 = (10000 * chan0.value) / (max_adc - chan0.value)
rchan1 = (10000 * chan1.value) / (max_adc - chan1.value)
rchan2 = (10000 * chan2.value) / (max_adc - chan2.value)
rchan3 = (10000 * chan3.value) / (max_adc - chan3.value)

vchan0 = (chan0.value / max_adc) * pga
vchan1 = (chan1.value / max_adc) * pga
vchan2 = (chan2.value / max_adc) * pga
vchan3 = (chan3.value / max_adc) * pga

# Create differential input between channel 0 and 1
# chan = AnalogIn(ads, ADS.P0, ADS.P1)

while True:
    print("   {:>5}\t{:>5}\t{:>5}\t{:>5}".format("raw", "v", "r", "rem_voltage"))
    print("A0 {:>5}\t{:>5.3f}\t{:>5}\t{:>5}".format(chan0.value, chan0.voltage, rchan0, vchan0))
    print("A1 {:>5}\t{:>5.3f}\t{:>5}\t{:>5}".format(chan1.value, chan1.voltage, rchan1, vchan1))
    print("A2 {:>5}\t{:>5.3f}\t{:>5}\t{:>5}".format(chan2.value, chan2.voltage, rchan2, vchan2))
    print("A3 {:>5}\t{:>5.3f}\t{:>5}\t{:>5}".format(chan3.value, chan3.voltage, rchan3, vchan3))
    print("")

    time.sleep(3.0)
~                    

```

and here is the output.

```
          raw      v	            r	                               rem_voltage
A0 20576	    2.574	4594.908441268423	0.6447709162127425
A1 32752	    4.094	10045.39320328794	1.026318868963829
A2 32752	    4.094	10045.39320328794	1.026318868963829
A3  4432	    0.550	725.0279109476587	0.13938282636636268

```

On A0 I have a thermistor connected. I don't have anything on A1, A2 or A3 for my test.  A2 is wired the same as A0  and A1 but I didn't show it in the drawings for simplicity. A3 is the only one wired differently because it will be connected to the 5V pressure sensor for the filter pressure readings. Again there is nothing connected to A3 for my test.

![image](https://user-images.githubusercontent.com/6274208/192154260-11bf05ff-b9c3-424c-95b3-79cfce9edbd4.png)




Without disconnecting anything, looking instead at REM, The values of A0, A1, and A2 are the same but A0 should not be because it actually has a sensor connected. It's almost as if the pool controller really isn't getting any input from the ADC why would that be? The Python code is running on the same raspberry pi.

![image](https://user-images.githubusercontent.com/6274208/192153918-e8cdf26c-d939-4785-a50c-fdc1ed15ff1a.png)


Finally looking at the thermistor table for a value of 4558 should result in a value somewhere between 111 and 113 Fahrenheit ( if I am reading the table correctly) but it actually around 76 degrees.

```
  {
    "input": 4561,
    "value": 111
  },
  {
    "input": 4367,
    "value": 113
  }
```

**Reply by rstrouse** [reply]
Neither ADC value of 20576 or 32767 is correct.  Are you sure you have the connections to the high voltage and low voltage side correct on the level shifter?

**Reply by rstrouse** [reply]
Also set the gain for the input to 4.096 and see if it maxes out or not.   Anything over 2.048 volts with that gain is supposed to max at 2.048 given the output from your sample code the gain appears to actually be set at 4.086 not 2.048 and the max_adc should be 2^15 on a single ended reading.

**Reply by Hect0rG** [reply]
I just checked it again. It appears to be connected correctly. 3.3v to LV, 5v to HV, Ground to both Gnd pins. I believe they are connected correctly. I have checked it a bunch of times. Could it be the level shifter I am using? https://amzn.to/3Rbw2GL

With the gain set to 4.096, this is the output.

![image](https://user-images.githubusercontent.com/6274208/192175375-7b330529-fffa-4244-8c0c-358554c03968.png)

(21,491 * 10000) / (65536 −21491) = 4879

Looking that up in the table would be about 108 Fahrenheit but again it's supposed to be reading ~76 

```
  {
    "input": 4981,
    "value": 107
  },
  {
    "input": 4766,
    "value": 109
  }
```

**Reply by rstrouse** [reply]
Take your multimeter and read the voltage at A0.

**Reply by rstrouse** [reply]
Actually your math is not correct.  The max for the ADS1115 is 2^15 or 32767 when reading single ended.  So 2.6865v represents 5373ohms

**Reply by tagyoureit** [reply]
@Hect0rG Also, your python code is set up for a 1015 chip while REM is setup for a 1115. The former is a 12-bit chip and the latter is a 16-bit chip.  This will definitely give you different values.

**Reply by tagyoureit** [reply]
Can we go back to your level shifter and source voltage for a minute?

Your level shifter needs to be supplied by both 3v3 and 5v.  Your red supply line (HV) should not be going to the ADS1115; it should be coming directly from the Pi.   You need to connect the HV on the level shifter directly to a 5v source and also connect the VDD on the ADS to [the same] 5v source.  If run a wire directly from 5v to HV your level shifter will have the right source voltage AND so will your ADS1115.

[Adafruit level shifter](https://www.adafruit.com/product/757)
![image](https://user-images.githubusercontent.com/1646840/192304422-b4254671-79f3-49c2-973e-9af7d8c648d1.png)

![image](https://user-images.githubusercontent.com/1646840/192305100-90fd5fa0-e449-407f-a0c4-bc7a93fb3104.png)

EDIT: it's hard to tell from your drawing. Maybe they are all connected?

**Reply by Hect0rG** [reply]
Yes it is connected as you show in your mark up. I have the 5V directly from the Pi connected to the Hv on the Level shifter. That same 5V pin also goes to the VDD on the ADC.


It looks like I was using that python script incorrectly. Thanks for pointing out the issue with the ADC above.  I did some more reading  [here](https://learn.adafruit.com/adafruit-4-channel-adc-breakouts/python-circuitpython) and updated the script. 

The updated script is this.

```
# SPDX-FileCopyrightText: 2021 ladyada for Adafruit Industries
# SPDX-License-Identifier: MIT

import time
import board
import busio
import adafruit_ads1x15.ads1115 as ADS
from adafruit_ads1x15.analog_in import AnalogIn

# Create the I2C bus
i2c = busio.I2C(board.SCL, board.SDA)

# Create the ADC object using the I2C bus
ads = ADS.ADS1115(i2c)

#set gain for PGA
#+-4.096 = 1
#+-2.048 = 2
#+-1.024 = 4

ads.gain = 2

# Create single-ended input on each channel
chan0 = AnalogIn(ads, ADS.P0)
chan1 = AnalogIn(ads, ADS.P1)
chan2 = AnalogIn(ads, ADS.P2)
chan3 = AnalogIn(ads, ADS.P3)

while True:
    print("   {:>5}\t{:>5}".format("raw", "v"))
    print("A0 {:>5}\t{:>5.3f}".format(chan0.value, chan0.voltage))
    print("A1 {:>5}\t{:>5.3f}".format(chan1.value, chan1.voltage))
"adc_test.py" 69 lines, 899036 characters
```

The raw value is always at max which matches what the REM is showing. 

With  ads.gain = 1 (+-4.096)
```
          raw	    v
A0      20631	2.580
A1      32767	4.096
A2      32767	4.096
A3       4401	0.556

```

With  ads.gain = 2 (+-2.048)
```
          raw	    v
A0      32767	2.048
A1      32767	2.048
A2      32767	2.048
A3       4221	0.261
```

Measuring with a voltmeter at the ADC
A0 = 2.599 volts (Thermistor connected)
A1 = 5.168 volts (Nothing connected)
A2 = 5.168 volts (Nothing connected)
A3 = 1.3 volts (Nothing connected)

**Reply by rstrouse** [reply]
A gain of 2.048 is not correct when you are connected to 5v.   The ADC will max out at the pga value and not record any voltages larger than that.  Given the numbers for A0 above that is 5157ohms or ~106F.   You can try setting the gain to 6.144v to see if the reference voltage above 5v is impacting it.

However, this appears to be a problem with 1 of 3 things.
1. The reference resistor is not reading 10k (check it with your multimeter)
2. The thermistor is not reading correctly (check the resistance with it disconnected then look at the table to see if the temp is correct)
3. The ADC settings at 5.2vdd need to have a reference setting to 6.144 or the ADC is defective.

**Reply by Hect0rG** [reply]
Ok I'll check those out as soon as I can and get back.  Thanks

**Reply by Hect0rG** [reply]
Ref resistor A0 = 9.93k
Ref resistor A1 = 9.91k
Ref resistor A2 = 9.90k
Thermistor resistance = 10.01k

with ADS = 2/3 (+-6.144)

```
             raw	    v
A0      13725	2.575
A1      27525	5.161
A2      27518	5.162
A3       2937	0.535
```

and finally, with a new ADC I see similar results

```
             raw	    v
A0      13810	2.590
A1      27429	5.146
A2      27441	5.145
A3       1385	0.259
```

**Comment by rstrouse** [comment]
You know 2.5v is ~74F.  These readings look correct and are probably the same on 4.096 gain.  I hooked up a level shifter and decided to duplicate your issue.  Turns out you don't really have one from a wiring standpoint.  So you must not have the feed or the probe definition correct.

We are measuring the voltage over the divider so set up your feed from the ADC as follows.  Choose ch1.voltage as the feed value and feed it to your probe.
![image](https://user-images.githubusercontent.com/47839015/192647332-f6e5ed8d-c1a8-47d9-b932-38ed794711e6.png)

Then on your probe definition set it up like this where the input type is voltage and the vccRef is 5v.
![image](https://user-images.githubusercontent.com/47839015/192647596-b1beacc6-4247-4965-92a9-1ad8f3a92304.png)

**Reply by Hect0rG** [reply]
@rstrouse thanks for your help. It is now working with my Air temp on my test system. I may try putting this back on the actual pool controller pi tomorrow.  As for what fixed it, I think my first issue was the incorrect wiring of the ADC and level shifter followed by incorrect settings. I also noticed my readings coming in and out at one point which ended up giving me intermittent good / bad readings which compounded my confusion. That ended up being a false contact to my sensor. At the end of the day I gained a much better understanding of how this all works so that's a plus. Thanks for all your help. 

What did you mean by "Turns out you don't really have one from a wiring standpoint."  when talking about the level shifter? 

Now for hopefully my final question. As you mentioned I noticed the gain of 4.096 and 6.144 both give me similar readings. Should I use one over the other?

--------------------------------------------------------------------------------------------------------------------------------------------

In case anyone runs into this in the future, This is the final setup when using 5V for your thermistors through an ADC and then a logic level shifter to shift it to a voltage the pi can handle. 

![image](https://user-images.githubusercontent.com/6274208/192697079-c0929af3-e08b-4c0b-8863-9c4c1efe3003.png)

It makes more sense to follow the order shown in this link to understand how it works but below are the quick and dirty pictures of how I got it to work. https://github.com/rstrouse/relayEquipmentManager/wiki/Temperature-Sensors I suggest reading this before so you get a better understanding. It took me some time to wrap my head around it. 

**General tab** should look like this. 

![gnome-shell-screenshot-TZ91S1](https://user-images.githubusercontent.com/6274208/192692627-f6b4dd59-23a8-493a-9f77-995d948b60e9.png)

**Generic Devices Tab**

![image](https://user-images.githubusercontent.com/6274208/192692294-77181ddc-02f0-4087-abab-ad8ae872ac9f.png)

**click on pencil**
![image](https://user-images.githubusercontent.com/6274208/192692390-21116072-cba6-40c4-9bbd-da1373ec7f86.png)

**click on feed tab** and add this feed ( This is the only sensor I am using at the moment) Air Sensor for ambient temp
![image](https://user-images.githubusercontent.com/6274208/192692501-17828837-c8d7-4616-bba8-747b46c8e4a2.png)

**I2C bus tab** Add the ADS1115 if not there.
![image](https://user-images.githubusercontent.com/6274208/192692952-a63ef7c3-8e0e-43c1-b6ea-2de9ce6a74aa.png)

Edit Channel 1 ( at the moment I am getting good readings with either 4.096 or 6.144 I opted for 4.096 )
![image](https://user-images.githubusercontent.com/6274208/192693067-08fa2402-4588-4b7e-a597-0ba71a74e037.png)

Finally, create the I2C feed
![image](https://user-images.githubusercontent.com/6274208/192693375-e33135a3-8fd2-40ef-9432-becff047d0ca.png)

You should see the Air Temp now populating. (Nevermind the pool temp in the pic, I was playing with a second thermistor)
![image](https://user-images.githubusercontent.com/6274208/192695673-78ebea29-91c5-407c-8271-0c6bc3477754.png)

**Comment by rstrouse** [comment]
> What did you mean by "Turns out you don't really have one from a wiring standpoint." when talking about the level shifter?

I should have done the math (correctly) to calculate the temperature at 2.575v.  Then I would not have had you run around and change the ADC.

> Now for hopefully my final question. As you mentioned I noticed the gain of 4.096 and 6.144 both give me similar readings. Should I use one over the other?

You can use either one but the 4.096 gain will give you slightly more resolution because at 4v you are way beyond frozen.

Congratulations on getting this working!

**Reply by Hect0rG** [reply]
Great thanks again! I appreciate you sticking with me while I struggled.  In the end while it took a while to get it going I learned a bit so no harm done. :)

**Comment by Telemachus1977** [comment]
I'm looking to use a similar transducer with a 3.3 volt adc.

Why is the level shifter needed at all?

Couldn't you just feed the transducer 3.3v and interpret the return values accordingly?

I read on stack exchange that these sensors have a transfer function:

Vout = Vcc(.66667 * P + .1)  with P in mPa

It's a linear function so should be easily interpretable, no?
